main.tf

```diff
@@ -451,10 +451,11 @@ resource "aws_efs_file_system" "this" {
 
 resource "aws_efs_mount_target" "this" {
   # we coalescelist in order to specify the resource keys when we create the subnets using the VPC or they're specified for us.  This works around the for_each value depends on attributes which can't be determined until apply error
-  for_each = {
-    for k, v in zipmap(coalescelist(var.private_subnets, var.private_subnet_ids), local.private_subnet_ids) : k => v
-    if var.enable_ephemeral_storage == false
-  }
+  for_each = (
+    var.enable_ephemeral_storage == false
```

Comment:
I don't think this change to the code is really necessary because it is identical to the existing one which works a bit better.

I can't find an exact issue in [Terraform core](https://github.com/hashicorp/terraform) but it is related to `map` != `object` and we have this approach in all other modules already.

Reply:
In its current form, it throws this error:
```
The "for_each" value depends on resource attributes that cannot be determined until apply, so Terraform cannot predict how many instances will be created. To work around this, use the -target argument to first apply only the resources that the for_each depends on.
```
So you can't apply the whole module, you need to somehow figure out which pieces should get applied first and apply those. Seems like a dealbreaker for using the module.

Reply:
I believe its because we are using `zipmap()` which converts computed resources into map keys and violates https://github.com/hashicorp/terraform/issues/4149

I think it would be better if we used `count` instead of `for_each` - we don't have a way of creating a static key for the `for_each` maps which means we'll always suffer against https://github.com/hashicorp/terraform/issues/4149

so line 454 becomes:

```hcl
  count = var.enable_ephemeral_storage == false ? local.private_subnet_ids : []
```

Reply:
It seems like you still need the `coalescelist`, though? Couldn't you do something like
```
for_each = (var.enable_ephemeral_storage == false ? toset(coalescelist(...)) : [])
```
Like that `efs_mount_target` resource isn't even using `each.key`, so it doesn't need to be a map?

Reply:
`for_each` here only accepts maps as input

Reply:
Ah, gotcha.

What's the best way to proceed here, then?

Reply:
@bryantbiggs @antonbabenko Checking in to see if I can help get this through.

Reply:
@greggilbert I think the best way would be if you could reproduce the problem (you could), fix the code, so that it works well for your and other cases.

Maybe simplification is necessary as @bryantbiggs mentioned about `zipmap()` but I don't think it is related.

The most important is to make it work for your case and the opposite one.

Sorry, that I can't provide a better solution without diving into this issue myself.